### DISCUSSION TITLE
All pkgs become "no bottle available!"

### DISCUSSION THREAD

**Discussion by vollkommenheit** [Homebrew/discussions#314] [Everyday usage] [answered]
After a "brew install [some-pkg]", which kicked off "brew update" first, I issued "cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core && git stash && git clean -d -f" (maybe something about no cleanup in the last 30 days?) as suggested by the eventual output, but the exact verbiage is now beyond my terminal's screen buffer. Thereafter, my homebrew seems to be hosed, as every pkg new or installed is reportedly "no bottle available!", even though the formulae are right there in my local /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula. After poking around based on internet searches, commands such as "brew repair" and "brew cleanup" used in salvage operations in the past seem no longer available. It appears to me that every time "brew update" gets triggered (I only recently found out that the auto "brew update" with every "brew install/upgrade/etc" could be turned off, but didn't bother), it pulls in the latest homebrew commits, so I'm wondering if any recent homebrew commit(s) might lead to such an issue.

**Comment by gromgit** [comment, marked as answer]
What do `brew config` and `brew doctor` output?

**Reply by vollkommenheit** [reply]
Nothing interesting at all in "brew config" other than it has the latest HEAD from the master branch tonight. 

"brew doctor" shows a bunch of unlinked dylibs, headers, .la, .a, .pc files, etc. But most of those are from software such as openssl-1.1 that I do long have built from sources directly, entirely outside of homebrew, also residing in /usr/local. So nothing interesting there either. In fact, I think these unlinked files resulted (tonight) as a result of the SNAFU with brew.

I am running an unsupported version of macOS, but it's been fine until tonight.

I could restore /usr/local/Homebrew from backup tomorrow, and am hopeful that would restore it back to "normal". But I'm curious what caused the breakage tonight, and if it's indeed as I suspect a result of one of the new brew commits from  auto "brew update" (after I ran "brew install some-pkg").

**Reply by carlocab** [reply]
You should probably share the output of `brew config` and `brew doctor` anyway. It contains a lot of info that's easy to miss; what looks uninteresting to you could actually be significant. (This actually happened to me earlier today: https://github.com/Homebrew/discussions/discussions/310#discussioncomment-221961)

**Reply by gromgit** [reply]
> I am running an unsupported version of macOS

That means the bottles for your OS version can be removed at any time, and without any notice whatsoever.

You can check this for yourself. Pick a formula that you're not getting bottles for, then `brew cat <formula>` and look at the `bottle` block therein. If you don't see an entry for your OS, it's been removed. If you *do* see an entry for your OS, then something's gone wrong in bottle storage, and you may wish to file an issue (though you might well get "unsupported OS" there too).

Either way, you're stuck with building from source until you upgrade macOS to a supported version, at which point "hey, what happened to all the bottles?" becomes a cause for everyone's concern.

_Sidenote:_ In future, when asking for technical help anywhere, please remember the Golden Rule: **Show, Don't Tell**. When folks trying to help ask you for specific output, they're expecting to see exactly that output, from which informed conclusions can be drawn and useful next steps suggested. They're _not_ expecting a (possibly inaccurate) summary of that output that might hide the real problem, along with (probably uninformed) conclusions that may well lead everyone on a wild goose chase.

**Reply by vollkommenheit** [reply]
@gromgit, completely understand. Looks like my unsupported version of macOS got removed from the formulae (at least for the pkgs I'm interested in). That explains it. So, would adding the relevant unsupported OS hashes back from the older formulae (from my backup) into the current formulae possibly work, with the caveat it's unsupported, or is there something more involved than that in the homebrew crew deciding to finally remove the unsupported OS versions from each individual formula (seemingly en masse recently) than that (since I notice the current OpenSSL.rb has a different list of supported OS from some other random formulae)?  What about I replace the entire current /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula with the one from my backup? What's the worst that could happen, other than the status quo (every pkg requires "--build-from-source"), with the understanding it's unsupported?

**Reply by vollkommenheit** [reply]
@carlocab, completely understand. That thread you linked reminds me that I did get something from one of the "brew" commands about possibly adding /usr/local/bin to my $PATH in ~/.bash_profile, even though I already have /usr/local/bin in front of /usr/bin in my ~/.bashrc, although it's not in the "brew doctor" output. In fact, I seem to recall a much shorter "brew doctor" output initially, after the brew formulae first broke last night.

**Reply by gromgit** [reply]
> would adding the relevant unsupported OS hashes back from the older formulae (from my backup) into the current formulae possibly work

If current formulae no longer list your OS, that almost certainly means they've been rebuilt with newer software versions or bug fixes, so inserting your old hashes into them won't work.

> the homebrew crew deciding to finally remove the unsupported OS versions from each individual formula

Both the maintainers and Homebrew CI infrastructure have more than enough stuff to deal with, so this is highly unlikely.

> I notice the current OpenSSL.rb has a different list of supported OS

Bottling for Apple's new M1 CPU is in progress. OpenSSL gets priority in the rebottling queue because it's a "foundational" formula -- you simply can't build many other formulae without it.

> What about I replace the entire current /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula with the one from my backup?

You'd have to restore the entire `/usr/local` hierarchy for this to work, since using current Homebrew software to manage old Homebrew formulae is almost as likely to fail as old software on current formulae. This is seriously **NOT** recommended, since:
* you'd never be able to update your software again
* the old bottles may disappear at any time (if they haven't already)
* you'll suffer whatever security issues would've been fixed by current versions of your desired software

**Comment by ivanmartcSourceM** [comment]
sorry i have used this command 
If -s is passed, scrub the cache, removing downloads for even the latest versions of formulae. Note downloads for any installed formulae will still not be deleted. If you want to delete those too: **rm -rf $(brew --cache)**

and i have deleted unixodbcc, now im not able to install it again using brew and can not install anything. pls help...  Mac big sur m1

**Reply by ivanmartcSourceM** [reply]
@ vitorgalvao

**Reply by carlocab** [reply]
What's the output when you install something? What is the output of `brew config` and `brew doctor`?

**Reply by ivanmartcSourceM** [reply]
the current problem output when try to reinstall or install something
`arch -x86_64 brew reinstall unixodbc`

<img width="585" alt="Screen Shot 2021-03-17 at 2 49 56 PM" src="https://user-images.githubusercontent.com/80718653/111529382-1676da80-8730-11eb-81f5-782ab6b25488.png">

the config output

<img width="643" alt="Screen Shot 2021-03-17 at 2 22 31 PM" src="https://user-images.githubusercontent.com/80718653/111529611-5c33a300-8730-11eb-8c79-84bb44dd1d9f.png">

the doctor output
<img width="604" alt="Screen Shot 2021-03-17 at 2 54 03 PM" src="https://user-images.githubusercontent.com/80718653/111529877-a9177980-8730-11eb-8f44-39df996a631f.png">

thanks for help

**Reply by carlocab** [reply]
Please [don't use screenshots](https://meta.stackoverflow.com/a/285557) when sharing the output of commands in your terminal. They're often difficult to read and impossible to reuse.

**Reply by ivanmartcSourceM** [reply]
but can you help me? pls...

**Reply by SMillerDev** [reply]
We will need the full text copied here to help, the images are not the complete information.

**Reply by ivanmartcSourceM** [reply]
got it wait...

**Comment by ivanmartcSourceM** [comment]


**Reply by carlocab** [reply]
Please [don't use screenshots](https://meta.stackoverflow.com/a/285557) when sharing the output of commands in your terminal. They're often difficult to read and impossible to reuse.

In this instance, your screenshots were not included in your message. Presumably, you were trying to attach them to your email, but that doesn't work with the GitHub UI.

**Comment by ivanmartcSourceM** [comment]
Thank you for the clarification.

**Comment by ivanmartcSourceM** [comment]
i have also used the brew cleanup function
